### PR TITLE
Expression Tree Java

### LINKED ISSUE #2051: Expression Tree- Java Implementaion
There is only C++ and python implementation of the expression tree. I want to add Java code for it.

### PR DESCRIPTION
**Pull request by RitikaJain8818**
Fixes #2051 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

#### Changes proposed in this pull request:
I have added the java code of expression code which was initially not there.

#### Languages Used:
Java

#### Files Added:
Added Expression_Tree.java

> We're happy to help you get this ready -- don't be afraid to ask for help.

Thanks!

### CODE REVIEW

**Comment by jainaman224** [comment]
Complete PR template. Add issue number to PR template after Fixes #.

**Comment by jainaman224** [comment]
Please complete the checklist.

File: Expression_Tree/Expression_Tree_Java.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
+import java.util.*;
+class Node { 
+
+	char value;
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree_Java.java]: Use 4 space indentation.

File: Expression_Tree/Expression_Tree_Java.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+/*
+sample input  : 395+2*+
+sample output : 3 + 9 + 5 * 2
+*/
+
+/*
+This code is contributed by Ritika Jain
+https://github.com/RitikaJain8818
+*/
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree_Java.java]: And 1 new line at the end of file.

File: Expression_Tree/Expression_Tree_Java.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	} 
+}
+
+/*
+sample input  : 395+2*+
+sample output : 3 + 9 + 5 * 2
+*/
+
+/*
+This code is contributed by Ritika Jain
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree_Java.java]: Remove this section.

File: Graph/HashMap_Implementation_Graph.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
+import java.util.ArrayList;
```
Comment (jainaman224) [inline Graph/HashMap_Implementation_Graph.java]: Remove this file. This PR should only contain code for Expression tree.

**Comment by RitikaJain8818** [comment]
Changes Done.
Please check and let me know!

**Comment by RitikaJain8818** [comment]
Thanks for these detailed suggested changes.
I have made the changes.
Please review my PRs.

**Comment by jainaman224** [comment]
Incorrect issue number.

File: Expression_Tree/Expression_Tree_Java.java [hunk 5]
```diff
@@ -0,0 +1,80 @@
+import java.util.*;
+
+class Node {
+
+    char value;
+	Node left, right;
+
+	Node(char data) {
+		value = data;
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree_Java.java]: Use 4 space indentation throughout the file.

File: Expression_Tree/Expression_Tree_Java.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
+import java.util.*;
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree_Java.java]: File name should be Expression_Tree.

File: Expression_Tree/Expression_Tree_Java.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+class Node {
+
+    char value;
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree_Java.java]: Inconsistent indentation.

File: Expression_Tree/Expression_Tree_Java.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		char[] charArray = postfix.toCharArray();
+		Node root = tree.constructTree(charArray);
+		System.out.print("Output: ");
+		tree.inorder_taversal(root);
+
+	}
+}
+
+/*
+ * sample input : 395+2*+ sample output : 3 + 9 + 5 * 2
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree_Java.java]: Move sample output to next line.

**Comment by RitikaJain8818** [comment]
I have made the suggested changes.
Please check and let me know.

File: Expression_Tree/Expression_Tree [hunk 9]
```diff
@@ -0,0 +1,81 @@
+import java.util.*;
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree]: File extension is absent.

File: Expression_Tree/Expression_Tree [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+class Node {
+
+    char value;
+	Node left, right;
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree]: Inconsistent indentation.

File: Expression_Tree/Expression_Tree [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+	Node(char data) {
+		value = data;
```
Comment (jainaman224) [inline Expression_Tree/Expression_Tree]: Use 4 space indentation.

**Comment by RitikaJain8818** [comment]
Changes Done.
Please check and let me know!

**Comment by RitikaJain8818** [comment]
It's been four days that I have made changes.
I haven't received any response.

File: Expression_Tree/Expression_Tree.java [hunk 12]
```diff
@@ -0,0 +1,81 @@
+import java.util.*;
+
+class Node {
+
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: Remove this line

File: Expression_Tree/Expression_Tree.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+    Node left, right;
+
+    Node(char data) {
+        value = data;
+        left = right = null;
+    }
+}
+
+class expressionTree {
+
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: remove this line

File: Expression_Tree/Expression_Tree.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+    public static void main(String args[]) {
+        Scanner scn = new Scanner(System.in);
+        String postfix = scn.next();
+        expressionTree tree = new expressionTree();
+        char[] charArray = postfix.toCharArray();
+        Node root = tree.constructTree(charArray);
+        System.out.print("Output: ");
+        tree.inorder_taversal(root);
+
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: remove this line

File: Expression_Tree/Expression_Tree.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+    void inorder_taversal(Node node) {
+        if (node != null) {
+            inorder_taversal(node.left);
+            System.out.print(node.value + " ");
+            inorder_taversal(node.right);
+        }
+    }
+
+    // func. to check if 'c' is an operator or not
+
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: remove this line

**Comment by somya-kapoor** [comment]
@RitikaJain8818 After making the changes, please squash the commits

**Comment by RitikaJain8818** [comment]
Changes are done.
Please check now.

File: Expression_Tree/Expression_Tree.java [hunk 16]
```diff
@@ -0,0 +1,65 @@
+import java.util.*;
+class Node {
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: Leave a line before it

File: Expression_Tree/Expression_Tree.java [hunk 17]
```diff
@@ -0,0 +1,65 @@
...
+    char value;
+    Node left, right;
+    Node(char data) {
+        value = data;
+        left = right = null;
+    }
+}
+class expressionTree {
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: Leave a line before it

File: Expression_Tree/Expression_Tree.java [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+        }
+    }
+    // func. to check if 'c' is an operator or not
+    boolean isOp(char cc) {
+        if (cc == '+' || cc == '-' || cc == '*' || cc == '/' || cc == '^') {
+            return true;
+        }
+        return false;
+    }
+    // Returns root of constructed tree for given postfix expression
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: Leave a line before it

File: Expression_Tree/Expression_Tree.java [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+}
+class expressionTree {
+    void inorder_taversal(Node node) {
+        if (node != null) {
+            inorder_taversal(node.left);
+            System.out.print(node.value + " ");
+            inorder_taversal(node.right);
+        }
+    }
+    // func. to check if 'c' is an operator or not
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: Leave a line before it

File: Expression_Tree/Expression_Tree.java [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+        if (cc == '+' || cc == '-' || cc == '*' || cc == '/' || cc == '^') {
+            return true;
+        }
+        return false;
+    }
+    // Returns root of constructed tree for given postfix expression
+    Node constructTree(char postfix[]) {
+        Stack<Node> stack = new Stack();
+        Node t, t1, t2;
+        // Traverse through each character of input expression
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: Leave a line before it

File: Expression_Tree/Expression_Tree.java [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+                // Pop two top nodes
+                t1 = stack.pop();
+                t2 = stack.pop();
+                // make children
+                t.right = t1;
+                t.left = t2;
+                stack.push(t);
+            }
+        }
+        // only element will be root of expression tree
```
Comment (somya-kapoor) [inline Expression_Tree/Expression_Tree.java]: Leave a line before it

**Comment by somya-kapoor** [comment]
@RitikaJain8818 Are you working on it?

**Comment by somya-kapoor** [comment]
@RitikaJain8818 Due to no activity for more than 7 days, I am closing this issue.

**Pull request closed without merging**